### PR TITLE
feat(x/ecocredit): implement basket MsgServer.Take

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/server/basket/take.go [hunk 1]
```diff
@@ -2,11 +2,185 @@ package basket
...
+				break
+			}
+
+			amountCreditsNeeded, err = amountCreditsNeeded.Sub(balance)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
```
Comment (technicallyty) [inline x/ecocredit/server/basket/take.go:152]: EventTake?

File: x/ecocredit/server/basket/take.go [hunk 2]
```diff
@@ -2,11 +2,185 @@ package basket
...
+	return &baskettypes.MsgTakeResponse{
+		Credits: credits,
+	}, nil
+}
+
+func (k Keeper) addCreditBalance(ctx context.Context, owner sdk.AccAddress, batchDenom string, amount math.Dec, retire bool, retirementLocation string) error {
+	sdkCtx := types.UnwrapSDKContext(ctx)
+	store := sdkCtx.KVStore(k.storeKey)
+	if !retire {
+		return ecocredit.AddAndSetDecimal(store, ecocredit.TradableBalanceKey(owner, ecocredit.BatchDenomT(batchDenom)), amount)
```
Comment (technicallyty) [inline x/ecocredit/server/basket/take.go:162]: EventReceive?

File: x/ecocredit/server/basket/take.go [hunk 3]
```diff
@@ -2,11 +2,185 @@ package basket
...
+	amountCreditsNeeded, err := amountBasketTokensDec.QuoExact(multiplier)
+	if err != nil {
+		return nil, err
+	}
+
+	var credits []*baskettypes.BasketCredit
+	for {
+		it, err := k.stateStore.BasketBalanceStore().List(ctx,
+			basketv1.BasketBalanceBasketIdBatchStartDateIndexKey{}.WithBasketId(basket.Id),
+		)
```
Comment (robert-zaremba) [inline x/ecocredit/server/basket/take.go:67]: shouldn't this be before the `for` loop?

File: x/ecocredit/server/basket/take_test.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+	k               basket.Keeper
+	ctrl            *gomock.Controller
+	acct            sdk.AccAddress
+	bankKeeper      *mocks.MockBankKeeper
+	ecocreditKeeper *mocks.MockEcocreditKeeper
+	fooBasketId     uint64
+	barBasketId     uint64
+	storeKey        *sdk.KVStoreKey
+	sdkCtx          sdk.Context
+}
```
Comment (robert-zaremba) [inline x/ecocredit/server/basket/take_test.go:46]: why not using https://pkg.go.dev/github.com/stretchr/testify/suite ?

File: x/ecocredit/server/basket/take_test.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+func setup(t *testing.T) *suite {
+	// prepare database
+	s := &suite{t: t}
+	var err error
+	s.db, err = ormdb.NewModuleDB(server.ModuleSchema, ormdb.ModuleDBOptions{})
```
Comment (robert-zaremba) [inline x/ecocredit/server/basket/take_test.go:52]: this err is not checked

File: x/ecocredit/server/basket/take_test.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	"github.com/cosmos/cosmos-sdk/orm/testing/ormtest"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/golang/mock/gomock"
+	basketv1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/basket/v1"
+	baskettypes "github.com/regen-network/regen-ledger/x/ecocredit/basket"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server/basket"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server/basket/mocks"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"gotest.tools/v3/assert"
```
Comment (robert-zaremba) [inline x/ecocredit/server/basket/take_test.go:29]: let's organize this imports

**Review by robert-zaremba — APPROVED** [review]
pre-approving

File: x/ecocredit/server/basket/take.go (hunk 3 above)
Comment (aaronc) [reply x/ecocredit/server/basket/take.go:67]: No we reopen the iterative each time. You cannot update/delete the item the iterative points to while it's open. Trying to allow mutable iterators upstream in the orm

**Review by technicallyty — APPROVED** [review]

File: x/ecocredit/server/basket/take_test.go (hunk 4 above)
Comment (aaronc) [reply x/ecocredit/server/basket/take_test.go:46]: Using gotest with orm and got used to it. It supported protobuf better

**Comment by aaronc** [comment]
@technicallyty can you take a look again and check that I got the events right. For `EventReceive`, I guess `sender` is `nil`??

**Comment by technicallyty** [comment]
> @technicallyty can you take a look again and check that I got the events right. For `EventReceive`, I guess `sender` is `nil`??

that's a good question.. maybe its something the interfaces team can answer? not sure what type of data they need/expect.

**Comment by aaronc** [comment]
And when we retire should there be both EventReceive and EventRetire?

**Comment by aaronc** [comment]
Maybe let's merge to unblock @robert-zaremba and then we'll add an events follow up

**Pull request merged**